pubsub/iterator.go

```diff
@@ -427,127 +437,111 @@ func (it *messageIterator) handleKeepAlives() {
+			it.pendingModAcks[id] = newSuccessAckResult()
 		}
 	}
 	it.checkDrained()
 }
 
-func (it *messageIterator) sendAck(m map[string]*AckResult) bool {
-	// Account for the Subscription field.
-	overhead := calcFieldSizeString(it.subName)
-	return it.sendAckIDRPC(m, maxPayload-overhead, func(ids []string) error {
-		recordStat(it.ctx, AckCount, int64(len(ids)))
-		addAcks(ids)
-		bo := gax.Backoff{
-			Initial:    100 * time.Millisecond,
-			Max:        time.Second,
-			Multiplier: 2,
-		}
-		cctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
-		defer cancel()
-		for {
-			// Use context.Background() as the call's context, not it.ctx. We don't
-			// want to cancel this RPC when the iterator is stopped.
-			cctx2, cancel2 := context.WithTimeout(context.Background(), 60*time.Second)
-			defer cancel2()
-			err := it.subc.Acknowledge(cctx2, &pb.AcknowledgeRequest{
-				Subscription: it.subName,
-				AckIds:       ids,
-			})
-			// Retry DeadlineExceeded errors a few times before giving up and
-			// allowing the message to expire and be redelivered.
-			// The underlying library handles other retries, currently only
-			// codes.Unavailable.
-			switch status.Code(err) {
-			case codes.DeadlineExceeded:
-				// Use the outer context with timeout here. Errors from gax, including
-				// context deadline exceeded should be transparent, as unacked messages
-				// will be redelivered.
-				if err := gax.Sleep(cctx, bo.Pause()); err != nil {
-					return nil
-				}
-			default:
-				// TODO(b/226593754): by default, errors should not be fatal unless exactly once is enabled
-				// since acks are "fire and forget". Once EOS feature is out, retry these errors
-				// if exactly-once is enabled, which can be determined from StreamingPull response.
-				return nil
+func (it *messageIterator) sendAck(m map[string]*AckResult) {
+	ackIDs := make([]string, 0, len(m))
+	for k := range m {
+		ackIDs = append(ackIDs, k)
+	}
+	var toSend []string
+	for len(ackIDs) > 0 {
+		toSend, ackIDs = splitRequestIDs(ackIDs, ackIDBatchSize)
+
+		recordStat(it.ctx, AckCount, int64(len(toSend)))
+		addAcks(toSend)
+		// Use context.Background() as the call's context, not it.ctx. We don't
+		// want to cancel this RPC when the iterator is stopped.
+		cctx2, cancel2 := context.WithTimeout(context.Background(), 60*time.Second)
+		defer cancel2()
+		err := it.subc.Acknowledge(cctx2, &pb.AcknowledgeRequest{
+			Subscription: it.subName,
+			AckIds:       toSend,
+		})
+		it.eoMu.RLock()
+		exactlyOnceDelivery := it.enableExactlyOnceDelivery
+		it.eoMu.RUnlock()
+		if exactlyOnceDelivery {
+			var md map[string]string
+			var st *status.Status
+			apiErr, ok := apierror.FromError(err)
+			if ok {
+				md = apiErr.Metadata()
+				st = apiErr.GRPCStatus()
 			}
+			resultsByAckID := make(map[string]*AckResult)
+			for _, ackID := range toSend {
+				resultsByAckID[ackID] = m[ackID]
+			}
+			// TODO(hongalex): retry the ackIDs with transient retriable errors.
```

Comment:
I'm assuming this is for the mentioned next PR?

Reply:
Yeah it is, I just left it as a todo for now.